feat: 添加数据库服务，数据库迁移模块以及添加新的数据实体

Bryasxin opened the pull request:
## 主要功能

### 1. 数据库实体

- **用户(Users)**：包含用户名、昵称和简介的用户管理
- **题目(Problems)**：包含标题、描述、时间/内存限制的编程题目
- **测试用例(TestCases)**：与题目关联的输入/输出测试用例
- **标签(Tags)**：可自定义的问题分类标签
- **题目标签(ProblemTag)**：题目和标签之间的多对多关系

### 2. 数据库服务

数据库服务采用 tarpc 框架实现，提供数据库操作的远程过程调用能力。

#### 用户操作

- 根据用户名获取用户信息
- 创建新用户
- 更新用户信息（昵称、简介）
- 删除用户（软删除）
- 检查用户是否存在
- 分页列出用户

#### 题目操作

- 根据标题获取题目
- 根据标签获取题目列表
- 根据标题或标签模式搜索题目
- 创建新题目
- 更新题目信息
- 删除题目（软删除）
- 分页列出题目

#### 测试用例操作

- 根据题目标题获取测试用例
- 创建单个或批量测试用例
- 更新测试用例
- 删除测试用例

#### 标签操作

- 根据名称获取标签
- 创建新标签
- 更新标签
- 删除标签
- 分页列出标签
- 根据模式搜索标签

### 3. 数据库迁移

包含了一套完整的迁移脚本来创建所有必要的表：

- 带有适当约束和索引的用户表
- 带有外键指向用户的题目表（创建者）
- 带有外键指向题目的测试用例表
- 用于问题分类的标签表
- 用于多对多关系的问题标签联结表
- 预填充的题目难度标签

### 4. 错误处理

为数据库操作提供了自定义错误类型：

- 数据库连接错误
- 实体未找到错误
- 无效参数错误
- 一般数据库错误

### 5. 软删除实现

为需要支持逻辑删除而非物理删除的实体实现了软删除特性：

- 在相关表中添加了 `deleted_at` 时间戳列
- 实现了处理软删除操作的特征(trait)

## 技术细节

### 架构

数据库服务采用客户端-服务器架构，使用 tarpc 进行通信。服务实现与接口定义分离，提高了模块化程度和可测试性。

### 依赖项

- sea-orm：用于数据库操作和 ORM 功能
- tarpc：用于 RPC 服务实现
- tracing：用于日志记录和监控
- chrono：用于时间相关操作

### 文件变更

1. **crates/database/**
   - `src/service.rs`：数据库服务特征定义
   - `src/service_impl.rs`：数据库操作的具体实现
   - `src/connection.rs`：数据库连接工具
   - `src/errors.rs`：自定义错误类型
   - `src/lib.rs`：公共 API 导出

2. **crates/entity/**
   - 所有数据库模型的实体定义
   - 软删除特征的实现

3. **migration/**
   - 创建所有表的迁移脚本
   - 难度标签的预填充

Bryasxin committed:
✨ feat(entity): 修改数据实体定义，添加软删除 trait，修改模块导出

- 使用 sea-orm 将原始数据实体改造为 sea-orm 的 Model
- 启用 sea-orm 依赖中 `macros` 和 `with-chrono` 的 feature
- 移除 user 数据定义中的 email 字段
- 添加 user 数据定义中的 deleted_at 和 is_active 字段
- 修改 user 数据定义中的 before_delete、before_save 和 new 行为
- 添加 user 数据定义中的 soft_delete 作为软删除
- 添加 SoftDelete 作为统一的数据实体需要实现的 trait

Bryasxin committed:
🔧 build: 添加开发和生产环境 docker-compose 文件

Bryasxin committed:
✨ feat: 初始化 database 模块

- 添加 `tokio` 依赖
- 添加 `tarpc` 依赖
- 创建 `database` 模块并新建接口与未完成的实现

Bryasxin committed:
✨ feat: 实现数据库微服务和用户实体序列化支持

1. 添加了数据库微服务实现：
    - 新增 service.rs 文件，实现了基于 tarpc 的数据库服务
    - 添加了 DatabaseService trait 和 DatabaseServiceImpl 实现
    - 支持通过用户名获取用户信息的远程调用方法
2. 更新了用户实体：
    - 为用户实体添加了 serde 序列化/反序列化支持
    - 可以在网络间传输用户数据
3. 完善了数据库连接管理：
    - 在 connection.rs 中实现了连接池和信号量控制
    - 添加了更完善的错误处理和连接管理逻辑
4. 配置文件更新：
    - 更新了 .env.example 文件，添加了数据库主机和端口配置
    - 修改了 docker-compose 文件，支持通过环境变量配置 PostgreSQL 端口
    - 更新了相关依赖配置

这些更改使数据库模块成为一个独立的微服务，可以通过 RPC 调用访问用户数据。

Bryasxin committed:
✨ feat(database): 添加用户资料更新功能

- 添加 `update_user_introduction` 和 `update_user_nickname` 接口更新用户资料

Bryasxin committed:
✨ feat: 添加 Tracing 日志系统

Bryasxin committed:
✨ feat: 添加 utilities 模块并重构日志初始化逻辑

Bryasxin committed:
✨ feat(database): 添加创建用户功能

Bryasxin committed:
✨ feat(database): 添加 chrono 依赖并重构错误处理

Bryasxin committed:
✨ feat(database): 添加数据库服务测试用例

Bryasxin committed:
🐞 fix(database): 修正测试仅在测试环境编译

Bryasxin committed:
✨ feat(database): 添加用户存在性检查和软删除功能

- 添加软删除和检查存在性方法
- 添加对用户名的检验
- 添加 todo 的测试用例

Bryasxin committed:
🐞 fix(database): 修正数据库更新方法的返回值

Bryasxin committed:
✨ feat(database): 完善数据库服务测试用例

为 `user_exists` 和 `get_user` 方法添加完整的测试逻辑

Bryasxin committed:
✨ feat(database): 优化数据库连接逻辑并添加错误处理

- 将注释修改为中文
- 添加连接数据库的 tracing 提示
- 连接数据库失败后使用 std::process 退出

Bryasxin committed:
✨ feat(migration): 添加用户表迁移和 CLI 设置

- 创建用户表的初始迁移
- 在 README 中包含所有必要的迁移命令文档
- 更新 .env.example 文件，添加数据库 URL 配置

Bryasxin committed:
🧪 test(database): 重构测试函数中服务实例的创建方式

- 移除函数 `create_service` 并转化为行内的方式

Bryasxin committed:
🧪 test(database): 完善测试用例

Bryasxin committed:
📃 docs(database): 修改 user 实体注释

Bryasxin committed:
🔧 build(workspace): 优化工作区配置

Bryasxin committed:
feat(database, utilities): 修改初始化日志系统工具函数

- 删除获取工作区环境变量作为模块日志等级
- 添加 `level_from_env` 可读取指定日志等级

Bryasxin committed:
🦄 refactor(database): 修改 user model 软删除方法

Bryasxin committed:
✨ feat(entity): 添加题目和测试用例实体模型

- 添加 `problem` 和 `test_case` 模块，包含字段定义、数据关联和软删除行为
- 完善 `users` 实体的软删除相关逻辑

Bryasxin committed:
✨ feat(migration): 添加问题和测试用例表的迁移模块

Bryasxin committed:
🐞 fix(entity): 完成添加对 problem 修改时的验证

Bryasxin committed:
🦄 refactor(entity): 修改 problem 实体的字段验证逻辑

Bryasxin committed:
✨ feat(entity): 新增标签和关联实体，更新关联关系

- 新增 `problem_tag` 和 `tag` 实体模块，用于支持题目与标签的多对多关联

Bryasxin committed:
✨ feat(migration): 添加标签表及关联表并修改问题表结构

Bryasxin committed:
✨ feat(migration): 合并迁移文件

Bryasxin committed:
✨ feat(migration): 添加题目难易度相关标签数据

Bryasxin committed:
🦄 refactor(database): 重构数据库服务接口并标记为 TODO，删除所有错误的测试

Bryasxin committed:
🦄 refactor(database): 将数据库服务实现分离到独立模块

Bryasxin committed:
✨ feat(entity): 为问题标题添加唯一性约束

Bryasxin committed:
🐳 chore(workspace): 更新 Cargo.lock 文件

Bryasxin committed:
✨ feat(database): 实现数据库服务核心功能

Bryasxin committed:
✨ feat(database): 添加获取 database 服务的客户端的方法

Bryasxin committed:
✨ feat(database): 简化数据库服务实现并添加问题验证逻辑

CN059 reviewed (commented):
代码没什么问题，组织良好，唯一的问题是，我现在要开发我们学校的公众号H5页面，前段时间没参与OJ的开发是因为在学Rust，过段时间我就赶快加入了，呜呜呜...你效率真的很高，呜呜呜

Bryasxin committed:
✨ feat(database, configuration): 新建 configuration crate 处理配置

- 将多个环境变量配置并移动到 yaml 配置文件
  - `POSTGRES_USER` -> `postgresql.user`
  - `POSTGRES_PASSWORD` -> `postgresql.password`
  - `POSTGRES_DB` -> `postgresql.database`
  - `POSTGRES_PORT` -> `postgresql.port`
  - `POSTGRES_HOST` -> `postgresql.host`
  - `DATABASE_SERVICE_SOCKET_ADDRESS` -> `services.database.addr`
- 新增多个数据库服务连接配置（services.database.connection）
- 引入 config crate 处理配置
- 定义多个配置结构体

Bryasxin committed:
✨ feat(database, config): 添加 sqlx 日志级别配置支持，改进数据库服务和数据库服务主函数实现

- 配置文件添加 `service.database.sqlx_logging_level` 读取 sqlx 日志等级
- 使用 `tokio::select!` 改进数据库服务主函数实现
- 改进数据库服务实现
- 连接池配置添加 sqlx_logging_level 配置
